postgresql: refactor and cleanup

wolfgangwalther opened the pull request:
## Description of changes

TLDR: This is a pure refactor, where `postgresql/default.nix` is split into multiple files and some `passthru` attributes are cleaned up. No derivations are changed.

Goal: I am currently working on building PG v16 via meson instead of autotools. This will make it possible to make `pkgsStatic.postgresql` work - at least for building `libpq`, which should cover almost all use-cases. This PR does not implement the meson build, yet. But to be able to have autotools-based builds up to v15 and meson-based builds for v16+ separated without repeating all the shared parts, this PR moves things around a little bit and splits them up in multiple files. But even without building with meson, the new structure is still an improvement, I think.

This is best reviewed commit-by-commit. I tried to make it reviewable despite the big chunks of code moved around. The most important fact: The derivations for all 5 versions with and without `_jit` don't change, so it's a pure refactor.

The files are structured like this:
- `default.nix` contains the logic to create all the top-level attributes, both with and without `_jit` suffix.
- `<major>.nix` files contain the version-specific stuff: version number, hashes, musl patches. This is recommended [here](https://github.com/NixOS/nixpkgs/tree/master/pkgs#versioning).
- `generic.nix` has all the metadata and most passthru arguments of the derivation, thus dealing with all the packages/extensions, tests and jit-helpers. It also contains all the patches, to make sure that both autotools and meson based builds will stay in sync.
- `autotools.nix` only has those parts which are specific to the autotools builds. This makes it easy to add a `meson.nix` next to it later.

Along the way, I also cleaned up a few things around arguments and passthru attributes:
- There was a `readline` passthru attribute introduced around 20 years ago - but entirely unused. Removed.
- `psqlSchema` is auto-created instead of passed for each version. Less chance for error when copy&pasting for a new version.
- `jitSupport` / `llvm` are not passed as passthru attributes to extensions, but via `callPackage` / the new pkgs scope. Again, reduces the chance for mistakes.
- The `this` argument is removed and replaced with internal recursion. This makes it possible to actually have things like `(postgresql.override { jitSupport = true; }).pkgs.postgis` work correctly without making sure to pass `this` all the time.

I also have commits to clean up `thisAttr` which suffers from the same problem for tests that `this` did - but those would not be a pure refactor anymore, so I'll put them in a separate PR.

Pinging maintainers: @thoughtpolice @danbst @globin @marsam @ivan @Ma27 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
postgresql: refactor to split up default.nix (2/2)

The recommended [1] structure for a package regarding versioning is to have each
version in a separate file. This commit just mechanically copies code around
without any changes.

Pure refactor, not changing any derivations.

[1]: pkgs/README.md

wolfgangwalther committed:
postgresql: refactor mkPackages in default.nix

Refactors some low hanging fruit in default.nix to make it easier to add new
versions later on.

Pure refactor, not changing any derivations.

This change makes it easier to add new versions in default.nix without messing
up - and also prevents us from adding version-specific arguments in default.nix
by accident in the future. Those should be put in the versioned .nix files
instead.

wolfgangwalther committed:
postgresql: refactor to move musl patches into versioned .nix files

This makes it obvious that the required argument muslPatches must be passed when
creating a new version file.

Pure refactor, not changing any derivations.

wolfgangwalther committed:
postgresql: refactor to move locale-binary-path.patch into patches/

Seems to have been put in the wrong place.

wolfgangwalther committed:
postgresql: refactor to remove "this" argument

This was proposed by abbradar in #150801, but left out of the follow up PR
#221851 by Ma27 to reduce the size of the diff. Compared to the initial
proposal this includes the callPackage call in the recursion, which avoids
breaking the withJIT/withoutJIT helpers.

In terms of nixpkgs, this is a pure refactor, no derivations change. However,
this makes downstream expressions like the following possible:

  (postgresql.override { jitSupport = true; }).pkgs.postgis

This would have not worked before without passing another "this" argument,
which is error prone as can be seen in this example:

  https://github.com/PostgREST/postgrest/pull/3222/files

wolfgangwalther committed:
postgresql: refactor to move packages.nix to ext/default.nix (2/2)

This aligns more with the commonly used style in nixpkgs.

wolfgangwalther committed:
postgresql: refactor to split up default.nix (1/2)

This just renames default.nix to generic.nix, because the biggest chunk
of code should move that way in the next commit. This gives us a much
better diff for the next commit and makes rebasing **much** easier in
case of changes. This commit does not stand on its own and needs to go
in with the next commit (2/2).

wolfgangwalther committed:
postgresql: refactor to remove passthru's readline attribute

This seems to have been introduced 20 years ago in 5863d4f - but
seems to have been a copy&paste mistake from the beginning.
AFAICT, it's not used anywhere.

wolfgangwalther committed:
postgresql: refactor to remove psqlSchema argument

The passthru attribute is still set, but automatically created from
the major version number. Fewer moving parts decrease the chance
for mistakes.

wolfgangwalther committed:
postgresql: refactor to remove useless references to "self"

No need to reference self here, because llvmPackages / stdenv' are available
in that scope anyway. Pure refactor, derivations don't change.

wolfgangwalther committed:
postgresql: refactor to pass jitSupport/llvm via scope instead of passthru

This makes it less error-prone to use the llvm package in extensions, because
it will always match the package used by the postgresql derivation itself.

Previously, you could've accidentally used llvm instead of postgresql.llvm
with a different result.

Ma27 commented:
May take until next week to get to it. But I'm definitely interested in reviewing that, thanks!

Ma27 commented:
Hmm, try running `nix-instantiate -A postgresql --argstr system x86_64-darwin`? Or whatever attribute is affected.

You don't need such hardware for the evaluation itself, that only gets relevant when it comes to building stuff, but the ofborg error looks like it didn't even get to that.

Ma27 commented:
Oh sorry I only saw the backtrace and didn't scroll down entirely :sweat_smile:

wolfgangwalther reviewed (commented):
Note: The latest push was just a rebase on latest master, no changes.

Ma27 commented on pkgs/servers/sql/postgresql/generic.nix:
```diff
@@ -0,0 +1,320 @@
+      };
+    in
+    {
+      psqlSchema = lib.versions.major version;
```
I'm wondering if we should still expose `jitSupport` here, that's the easiest way for other packages to check if postgresql is built with or without that.

Ma27 replied:
So, the way I thought about this was: we effectively have two variants of postgresql packaged, JIT and non-JIT (and multiple releases, but those can be found out via `.version`) and I think it doesn't hurt to expose the (rather prominent) detail of the package.

One thing I could imagine is a reusable module that expects to have a JIT variant for $service and asserts against that. It's not the reponsibility of that module to set the postresql package if it's for a different service and I don't think you want that because you'll run into issues quickly then when you'll need to patch the package on your own.

Ma27 replied:
> But JIT support doesn't change PostgreSQL functionally, afaik

To my knowledge, yes. But modules warning in case of unsuitable configurations is still a legit use-case (and I can certainly imagine applications where you want that).

> Personally, I could see more benefit in exposing enableSystemd or enablePython

I think you're right.
Postgresql extensions have a nice way to discover that and if you really need it you can still use `hasInfix "-jit-"`.

The core problem behind this is that we just don't have a good configuration story for packages (as in, you can read from `config.services.foo` how service `foo` is configured, but you can't do the same for `pkgs.foo` out of the box), so this should be solved at a different level.

> btw, just noting: Should we maybe rename jitSupport to enableJIT for consistency

Yes, but I'd suggest to leave `jitSupport ? null` around for 1 release and raise a warning if it's not null (i.e. `true`/`false`). `enableJIT` should be set to `jitSupport` in that case, then. Sure, NixOS users won't even notice because they'll probably use the module, but I'm pretty sure there are a few nixpkgs users out there.

These package arguments not only have horrible discovery, but as a result also horrible (i.e. non-existant) ways to communicate changes (short of "This attribute isn't supposed to be part of this function argument").

Ma27 commented on nixos/modules/services/databases/postgresql.md:
````diff
@@ -277,7 +277,7 @@ self: super: {
 ```
 self: super: {
-  postgresql_15 = super.postgresql_15.override { this = self.postgresql_15; } // {
+  postgresql_15 = super.postgresql_15.override {} // {
````
The `.override {}` shouldn't be needed anymore here.

Ma27 commented on pkgs/servers/sql/postgresql/generic.nix:
```diff
@@ -0,0 +1,320 @@
+      })
+
+    ] ++ lib.optionals stdenv'.hostPlatform.isMusl (
+      map fetchurl (lib.attrValues muslPatches)
```
I'm wondering why this was `fetchurl` in the first place, maybe we should switch back to `fetchpatch`. You should check if the hashes are different then.
If so, this needs to go to a follow-up PR.

marsam replied:
`fetchurl` was used because those patches are static files from AlpineLinux. On the other hand, `fetchpatch` is recommended when the patch is generated dynamically https://github.com/NixOS/nixpkgs/issues/240141
Nonetheless, I'm not against changing it to `fetchpatch`

Ma27 commented:
Hmm... would you mind trying out if adding 2da789a409c67c99c7e7b27730256d42f10a5601 to `.git-blame-ignore-revs` helps to retain a bit of the history of the previous file?

No idea if that works out, but maybe worth a try?

Anyways, I like that change and the fact that we got a few cleanups ready without any rebuilds.

Will let a few other people review as well before merging. cc @marsam @thoughtpolice @ivan

wolfgangwalther committed:
postgresql: refactor to move packages.nix to ext/default.nix (1/2)

This commit is split up into two commits to allow git to detect renames,
make rebasing easier and allow a working entry in .git-blame-ignore-revs.

To allow bisecting we allow evaluation on every commit by moving the extensions
into ext/ext/ first and back to ext/ with the next commit.

Ma27 commented:
> I think that's a great idea, not sure whether to wait to do this in a second PR (maybe in the follow up cleanup PR)

Fair.

Ma27 commented:
> To make this work, we will have to split the commit into two commits

Hmm, there's really no other way to trick git into detecting the change otherwise? :/ 

I mean, we're trading the ability to bisect postgresql here.
That said, I think it should be possible to fix the issue with the other rename by `callPackage`ing `postgresql/generic.nix` in the first commit, right?

wolfgangwalther committed:
postgresql: refactor to remove doInstallCheck = false

This is the default anyway.

wolfgangwalther committed:
postgresql: rename enableSystemd to systemdSupport for consistency

We have gssSupport, jitSupport and pythonSupport - but enableSystemd?

Across nixpkgs there are currently three styles commonly used, about equally
often: withX, xSupport and enableX.

Let's at least use one consistent style in this package.

Ma27 commented:
> I think this is in good shape

Agreed.

Ping @ajs124 @marsam @yu-re-ka you were the last folks who touched the package itself.

I'd consider merging it at some point next week though, I'd rather not stall this.

wolfgangwalther committed:
postgresql: explain use of fetchurl for alpine patches

marsam reviewed (approved):
LGTM. Good work, thank you!

Ma27 reviewed (approved):
Great work!
